pypit/arproc.py

```diff
@@ -604,62 +637,124 @@ def get_ampscale(slf, det, msflat):
 
 def get_datasec_trimmed(slf, fitsdict, det, scidx):
     """
-     Generate a frame that identifies each pixel to an amplifier, and then trim it to the data sections.
-     This frame can be used to later identify which trimmed pixels correspond to which amplifier
+    Primarily a wrapper with calls to get_datasec and pix_to_amp()
 
     Parameters
     ----------
-    slf : class
-      An instance of the ScienceExposure class
-    fitsdict : dict
-      Contains relevant information from fits header files
+    slf
+    fitsdict
     det : int
-      Detector number, starts at 1
     scidx : int
-      Index of science frame
 
     Returns
     -------
-    fitsdict : dict
-      Updates to the input fitsdict
+    Fills slf._datasect
+    fitsdict['naxis0'] and fittsdict['naxis1']
     """
     dnum = settings.get_dnum(det)
+    spectrograph = settings.argflag['run']['spectrograph']
+    scifile = fitsdict['directory'][scidx]+fitsdict['filename'][scidx]
+    numamplifiers = settings.spect[dnum]['numamplifiers']
+
+    # Instrument specific bits
+    if spectrograph in ['keck_lris_blue', 'keck_lris_red', 'keck_deimos']:
+        # Grab
+        datasec, oscansec, naxis0, naxis1 = get_datasec(spectrograph, scifile,
+                                                        numamplifiers=numamplifiers, det=det)
+        # Fill (for backwards compatability)
+        for kk in range(numamplifiers):
+            sdatasec = "datasec{0:02d}".format(kk+1)
+            settings.spect[dnum][sdatasec] = datasec[kk]
+            soscansec = "oscansec{0:02d}".format(kk+1)
+            settings.spect[dnum][soscansec] = oscansec[kk]
+        fitsdict['naxis0'][scidx] = naxis0
+        fitsdict['naxis1'][scidx] = naxis1
+
+    # Build the datasec lists for pix_to_amp
+    datasec = []
+    for i in range(numamplifiers):
+        sdatasec = "datasec{0:02d}".format(i+1)
+        datasec.append(settings.spect[dnum][sdatasec])
+    # Call
+    naxis0, naxis1 = int(fitsdict['naxis0'][scidx]), int(fitsdict['naxis1'][scidx])
+    slf._datasec[det-1] = pix_to_amp(naxis0, naxis1, datasec, numamplifiers)
+    return
+
+
+def get_datasec(spectrograph, scifile, numamplifiers=None, det=None):
+    """  Determine the data and overscan sections of an image
+
+    Currently only used for LRIS and DEIMOS (with their multiple detectors
+    packed in funny ways).  Should consider another approach.
+
+    Parameters
+    ----------
+    spectrograph : str
+    scifile : str
+    numamplifiers : int (optional)
+    det : int (optional)
+      Detector number, starts at 1
 
+    Returns
+    -------
+    fitsdict : dict
+      Updates to the input fitsdict
+    """
     # Get naxis0, naxis1, datasec, oscansec, ampsec for specific instruments
-    if settings.argflag['run']['spectrograph'] in ['keck_lris_blue', 'keck_lris_red']:
+    datasec, oscansec, naxis0, naxis1 = [], [], 0, 0
+    if spectrograph in ['keck_lris_blue', 'keck_lris_red']:
```

Comment:
Continuing from the last statement...  The reason to pull it higher is that it will make the inclusion of new/user instruments easier/possible.

Reply:
agreed